packages/core/src/modules/drpc/DRPCMessagesApi.ts

```diff
@@ -0,0 +1,173 @@
+    if (!isValidDRPCRequestObject(message)) {
+      return {
+        jsonrpc: '2.0',
+        id: message.id,
+        error: { code: DRPCErrorCode.INVALID_REQUEST, message: 'Invalid request' },
+      }
+    }
+    const handler = this.drpcMethodHandlers.get(message.method)
+    if (!handler) {
+      return {
+        jsonrpc: '2.0',
+        id: message.id,
+        error: { code: DRPCErrorCode.METHOD_NOT_FOUND, message: 'Method not found' },
+      }
+    }
+
+    try {
+      let response: DRPCResponseObject | Record<string, never> = await handler(message)
+      if (message.id === null) {
+        response = {}
+      }
+      return response
+    } catch (error) {
+      return {
+        jsonrpc: '2.0',
+        id: message.id,
+        error: { code: DRPCErrorCode.INTERNAL_ERROR, message: 'Internal error', data: error.message },
+      }
+    }
+  }
+
+  public async sendDRPCRequest(connectionId: string, request: DRPCRequest): Promise<DRPCResponse> {
+    const messageId = uuid()
+    const connection = await this.connectionService.getById(this.agentContext, connectionId)
+    try {
+      const { message: drpcMessage, record: drpcMessageRecord } = await this.drpcMessageService.createRequestMessage(
+        this.agentContext,
+        request,
+        connection,
+        messageId
+      )
+      await this.sendDRPCMessage(connection, drpcMessage, drpcMessageRecord)
+    } catch {
+      throw new Error('Invalid DRPC Request')
+    }
+    return new Promise((resolve) => {
+      const listener = ({
+        message,
+        removeListener,
+      }: {
+        message: DRPCRequestMessage | DRPCResponseMessage
+        removeListener: () => void
+      }) => {
+        if (message instanceof DRPCResponseMessage && message.threadId === messageId) {
+          removeListener()
+
+          resolve(message.response)
+        }
+      }
+
+      this.drpcMessageService.createMessageListener(listener)
+    })
+  }
+
+  private async sendDRPCResponse(connectionId: string, threadId: string, response: DRPCResponse): Promise<void> {
+    const connection = await this.connectionService.getById(this.agentContext, connectionId)
+    const { message: drpcMessage, record: drpcMessageRecord } = await this.drpcMessageService.createResponseMessage(
+      this.agentContext,
+      response,
+      connection
+    )
+    drpcMessage.setThread({ threadId })
+    await this.sendDRPCMessage(connection, drpcMessage, drpcMessageRecord)
+  }
+
+  private async sendDRPCMessage(
+    connection: ConnectionRecord,
+    message: DRPCRequestMessage | DRPCResponseMessage,
+    messageRecord: DRPCMessageRecord
+  ): Promise<void> {
+    const outboundMessageContext = new OutboundMessageContext(message, {
+      agentContext: this.agentContext,
+      connection,
+      associatedRecord: messageRecord,
+    })
+
+    await this.messageSender.sendMessage(outboundMessageContext)
+  }
+
+  public createDRPCMethodHandler(
```

Comment:
```suggestion
  public addMethodHandler(
```

Reply:
Probably would be nice to add some validation if a handler for a given method is already registered, as only one per method is allowed, right?

Reply:
I'm thinking whether a handler should be registered at all in Credo? 

I think the normal event flow may work well for this:
- receive rpc message, get notified using agent.events.on
- execute code as part of handler, this is not handled by credo
- call api to send response.

Reply:
Especially with multi tenancy the add handler may not be the best fit (also think about multi-instance deployments). So i'd prefer to have just the message sending/receiving handled by credo and the actual procedure be executed based on events

Reply:
I didn't even consider the aspect of multi tenancy so that's also a very good point :).